datacube/api/core.py

```diff
@@ -46,53 +47,103 @@ class Datacube(object):
+            contain a 'default' section.
 
             Allows you to have multiple datacube instances in one configuration, specified on load,
             eg. 'dev', 'test' or 'landsat', 'modis' etc.
 
-        :param bool validate_connection: Should we check that the database connection is available and valid
+            If env is an ODCEnvironment object, config and index should both None.
+
+        :param raw_config: Explicit configuration to use.  Either as a string (serialised in ini or yaml format) or
+            a dictionary (deserialised).  If provided, config should be None.
+
+        :param bool validate_connection: Should we check that the database connection is available and valid.
+            Defaults to True. Ignored if index is passed.
 
         :return: Datacube object
 
         """
 
-        def normalise_config(config):
-            if config is None:
-                return LocalConfig.find(env=env)
-            if isinstance(config, str):
-                return LocalConfig.find([config], env=env)
-            return config
+        # Validate arguments
+
+        if index is not None:
+            # If an explicit index is provided, all other index-creation arguments should be None.
+            should_be_none: list[str] = []
+            if config is not None:
+                should_be_none.append["config"]
+            if raw_config is not None:
+                should_be_none.append["raw_config"]
+            if app is not None:
+                should_be_none.append["app"]
+            if env is not None:
+                should_be_none.append["env"]
+            if should_be_none:
+                raise ValueError(
+                    f"When an explicit index is provided, these arguments should be None: {','.join(should_be_none)}"
+                )
+
+        if config is not None and raw_config is not None:
+            # Cannot supply both config (file paths) and raw_config (explicit text/dictionary)
+            raise ValueError("Cannot supply both the 'config' and 'raw_config' arguments")
+
+        if config is not None or raw_config is not None:
+            # if config or raw_config provided, env cannot be an ODCEnvironment
+            if isinstance(env, ODCEnvironment):
+                raise ValueError(
+                    "If env is passed as an ODCEnvironment object, config and raw_config must both be None"
+                )
 
-        if index is None:
-            index = index_connect(normalise_config(config),
-                                  application_name=app,
-                                  validate_connection=validate_connection)
+        if index is not None:
+            # Explicit index passed in?  Use it.
+            self.index = index
+            return
+
+        # Obtain an ODCEnvironment object:
+        if isinstance(env, ODCEnvironment):
+            cfg_env = env
+        elif isinstance(config, ODCConfig):
```

Comment:
Would it perhaps be cleaner to have a single `config` parameter in here, and then ascertain type (and handling) in the `ODCConfig` class? I wonder if the extra number of params this introduces in both classes might make things a little confusing

Reply:
I think it's just the price of having the concept of environments within larger configurations.

Reply:
I moved this logic to the ODCConfig.get_environment() class variable.